## Homework

In this homework, we will use Credit Card Data from [the previous homework](https://github.com/alexeygrigorev/mlbookcamp-code/blob/master/course-zoomcamp/cohorts/2022/04-evaluation/homework.md).

## Question 1

* Install Pipenv
* What's the version of pipenv you installed?
    - __pipenv, version 2022.9.24__
* Use `--version` to find out

In [1]:
!pip install pipenv

/home/elite/.local/lib/python3.8/site-packages/pkg_resources/__init__.py:123: PkgResourcesDeprecationWarning: 0.1.36ubuntu1 is an invalid version and will not be supported in a future release
  warnings.warn(
/home/elite/.local/lib/python3.8/site-packages/pkg_resources/__init__.py:123: PkgResourcesDeprecationWarning: 0.23ubuntu1 is an invalid version and will not be supported in a future release
  warnings.warn(
     |████████████████████████████████| 3.3 MB 1.6 MB/s eta 0:00:01
     |████████████████████████████████| 8.8 MB 9.2 MB/s eta 0:00:01
     |████████████████████████████████| 468 kB 10.9 MB/s eta 0:00:01


In [3]:
!pipenv --version

pipenv, version 2022.9.24


## Question 2

* Use Pipenv to install Scikit-Learn version 1.0.2
* What's the first hash for scikit-learn you get in Pipfile.lock?

__08ef968f6b72033c16c479c966bf37ccd49b06ea91b765e1cc27afefe723920b__

In [5]:
!pipenv install scikit-learn==1.0.2

Creating a virtualenv for this project...
Pipfile: /home/elite/Documents/ML-Zoomcamp-2022/homework/Pipfile
Using /usr/bin/python3 (3.8.10) to create virtualenv...
⠋ Creating virtual environment...created virtual environment CPython3.8.10.final.0-64 in 688ms
  creator CPython3Posix(dest=/home/elite/.local/share/virtualenvs/homework-1YVHYzkr, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/elite/.local/share/virtualenv)
    added seed packages: pip==22.2.2, setuptools==65.3.0, wheel==0.37.1
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator

✔ Successfully created virtual environment! 
Virtualenv location: /home/elite/.local/share/virtualenvs/homework-1YVHYzkr
Creating a Pipfile for this project...
Installing scikit-learn==1.0.2...
Adding scikit-learn to Pipfile's [packages]...
✔ Installation Succeeded 
Pipfile.lock not 

In [21]:
!pipenv shell

Launching subshell in virtual environment...
]0;elite@the-python-ninja: ~/Documents/ML-Zoomcamp-2022/homeworkelite@the-python-ninja:~/Documents/ML-Zoomcamp-2022/homework$  . /home/elite/.local/share/virtualenvs/homework-1YVHYzkr/bin/activate
(homework) ]0;elite@the-python-ninja: ~/Documents/ML-Zoomcamp-2022/homeworkelite@the-python-ninja:~/Documents/ML-Zoomcamp-2022/homework$ ^C

(homework) ]0;elite@the-python-ninja: ~/Documents/ML-Zoomcamp-2022/homeworkelite@the-python-ninja:~/Documents/ML-Zoomcamp-2022/homework$ 

## Models

We've prepared a dictionary vectorizer and a model.

They were trained (roughly) using this code:

```python
features = ['reports', 'share', 'expenditure', 'owner']
dicts = df[features].to_dict(orient='records')

dv = DictVectorizer(sparse=False)
X = dv.fit_transform(dicts)

model = LogisticRegression(solver='liblinear').fit(X, y)
```

> **Note**: You don't need to train the model. This code is just for your reference.

And then saved with Pickle. Download them:

* [DictVectorizer](https://github.com/alexeygrigorev/mlbookcamp-code/blob/master/course-zoomcamp/cohorts/2022/05-deployment/homework/dv.bin?raw=true)
* [LogisticRegression](https://github.com/alexeygrigorev/mlbookcamp-code/blob/master/course-zoomcamp/cohorts/2022/05-deployment/homework/model1.bin?raw=true)

With `wget`:

```bash
PREFIX=https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/course-zoomcamp/cohorts/2022/05-deployment/homework
wget $PREFIX/model1.bin
wget $PREFIX/dv.bin
```

In [14]:
PREFIX='https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/course-zoomcamp/cohorts/2022/05-deployment/homework'
!wget $PREFIX/model1.bin
!wget $PREFIX/dv.bin

--2022-10-11 15:46:34--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/course-zoomcamp/cohorts/2022/05-deployment/homework/model1.bin
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 889 [application/octet-stream]
Saving to: ‘model1.bin’

model1.bin          100%[===================>]     889  --.-KB/s    in 0s      

2022-10-11 15:46:35 (13.4 MB/s) - ‘model1.bin’ saved [889/889]

--2022-10-11 15:46:35--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/course-zoomcamp/cohorts/2022/05-deployment/homework/dv.bin
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.1

## Question 3

Let's use these models!

* Write a script for loading these models with pickle
* Score this client:

```json
{"reports": 0, "share": 0.001694, "expenditure": 0.12, "owner": "yes"}
```

What's the probability that this client will get a credit card? 

* __0.148__ - Closest to 0.1621
* 0.391
* 0.601
* 0.993

If you're getting errors when unpickling the files, check their checksum:

```bash
$ md5sum model1.bin dv.bin
3f57f3ebfdf57a9e1368dcd0f28a4a14  model1.bin
6b7cded86a52af7e81859647fa3a5c2e  dv.bin
```

In [20]:
%%writefile week_5.py
import pickle
import sklearn

model_file = 'model1.bin'
dv_file = 'dv.bin'

client = {"reports": 0, "share": 0.001694, "expenditure": 0.12, "owner": "yes"}

with open(model_file, 'rb') as f:
    model = pickle.load(f)
    
with open(dv_file, 'rb') as f:
    dict_vect = pickle.load(f)
    
X = dict_vect.transform(client)

prediction = model.predict_proba(X)[0,1]
print(f"Probability of client getting a credit card: {prediction}")

Probability of client getting a credit card: 0.16213414434326598


/home/elite/.local/lib/python3.8/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator LogisticRegression from version 1.0.2 when using version 1.1.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
  warnings.warn(
/home/elite/.local/lib/python3.8/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator DictVectorizer from version 1.0.2 when using version 1.1.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
  warnings.warn(


In [12]:
!md5sum model1.bin dv.bin

3f57f3ebfdf57a9e1368dcd0f28a4a14  model1.bin
6b7cded86a52af7e81859647fa3a5c2e  dv.bin


## Question 4

Now let's serve this model as a web service

* Install Flask and gunicorn (or waitress, if you're on Windows)
* Write Flask code for serving the model
* Now score this client using `requests`:

```python
url = "YOUR_URL"
client = {"reports": 0, "share": 0.245, "expenditure": 3.438, "owner": "yes"}
requests.post(url, json=client).json()
```

What's the probability that this client will get a credit card?

* 0.274
* 0.484
* 0.698
* __0.928__

In [22]:
!pipenv install flask gunicorn

Installing flask...
Adding flask to Pipfile's [packages]...
✔ Installation Succeeded 
Installing gunicorn...
Adding gunicorn to Pipfile's [packages]...
✔ Installation Succeeded 
Pipfile.lock (3db960) out of date, updating to (b0a961)...
Locking [packages] dependencies...
Building requirements...
Resolving dependencies...
✔ Success! 
Locking [dev-packages] dependencies...
Updated Pipfile.lock (b0a961)!
Installing dependencies from Pipfile.lock (b0a961)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.


In [30]:
%%writefile serve_model.py

from flask import Flask, request, jsonify
import pickle


model_file = 'model1.bin'
dv_file = 'dv.bin'


with open(model_file, 'rb') as f:
    model = pickle.load(f)
    
with open(dv_file, 'rb') as f:
    dict_vect = pickle.load(f)

app = Flask('credit_card')

@app.route('/predict', methods=['POST'])
def predict():
    
    client = request.get_json()
    X_test = dict_vect.transform(client)
    y_pred_proba = model.predict_proba(X_test)[:,1]
    result = {
        'card_approved_probability': float(y_pred_proba)
    }

    return jsonify(result)

if __name__ == '__main__':
    app.run(debug=True, host='0.0.0.0')



Overwriting serve_model.py


In [31]:
import requests
url = "http://127.0.0.1:5000/predict"
client = {"reports": 0, "share": 0.245, "expenditure": 3.438, "owner": "yes"}
response = requests.post(url, json=client).json()
print(response)


{'card_approved_probability': 0.9282218018527452}


## Docker

Install [Docker](https://github.com/alexeygrigorev/mlbookcamp-code/blob/master/course-zoomcamp/05-deployment/06-docker.md). We will use it for the next two questions.

For these questions, we prepared a base image: `svizor/zoomcamp-model:3.9.12-slim`. 
You'll need to use it (see Question 5 for an example).

This image is based on `python:3.9.12-slim` and has a logistic regression model 
(a different one) as well a dictionary vectorizer inside. 

This is how the Dockerfile for this image looks like:

```docker 
FROM python:3.9.12-slim
WORKDIR /app
COPY ["model2.bin", "dv.bin", "./"]
```

We already built it and then pushed it to [`svizor/zoomcamp-model:3.9.12-slim`](https://hub.docker.com/r/svizor/zoomcamp-model).

> **Note**: You don't need to build this docker image, it's just for your reference.

## Question 5

Download the base image `svizor/zoomcamp-model:3.9.12-slim`. You can easily make it by using [docker pull](https://docs.docker.com/engine/reference/commandline/pull/) command.

So what's the size of this base image?

* 15 Mb
* __125 Mb__
* 275 Mb
* 415 Mb

You can get this information when running `docker images` - it'll be in the "SIZE" column.

In [32]:
!docker pull svizor/zoomcamp-model:3.9.12-slim

3.9.12-slim: Pulling from svizor/zoomcamp-model

a5fb9032: Pulling fs layer 
81b69b9a: Pulling fs layer 
a0a62ab5: Pulling fs layer 
28511179: Pulling fs layer 
79bc1c47: Pulling fs layer 
c3f7336a: Pulling fs layer 
Digest: sha256:10445b40653d5ac17ede84db17f42ae8c4090b347a979372b8102174498b33b9[6A
Status: Downloaded newer image for svizor/zoomcamp-model:3.9.12-slim
docker.io/svizor/zoomcamp-model:3.9.12-slim


In [33]:
!docker images

REPOSITORY              TAG           IMAGE ID       CREATED        SIZE
svizor/zoomcamp-model   3.9.12-slim   571a6fdc554b   9 days ago     125MB
localstack/localstack   latest        8d85856e4e90   2 months ago   1.44GB


## Dockerfile

Now create your own Dockerfile based on the image we prepared.

It should start like that:

```docker
FROM svizor/zoomcamp-model:3.9.12-slim
# add your stuff here
```

Now complete it:

* Install all the dependencies form the Pipenv file
* Copy your Flask script
* Run it with Gunicorn 

After that, you can build your docker image.

In [60]:
%%writefile Dockerfile

FROM svizor/zoomcamp-model:3.9.12-slim

RUN pip install pipenv

WORKDIR /homework

COPY ["Pipfile", "Pipfile.lock", "./"]

RUN pipenv install --system --deploy

COPY ["serve_model.py", "model1.bin", "dv.bin", "./"]

EXPOSE 5000

ENTRYPOINT ["python3"]
CMD [ "serve_model.py", "--host=0.0.0.0"]


Overwriting Dockerfile


In [61]:
!docker build -t svizor/zoomcamp-model:predict .

Sending build context to Docker daemon  1.976MB
Step 1/9 : FROM svizor/zoomcamp-model:3.9.12-slim
 ---> 571a6fdc554b
Step 2/9 : RUN pip install pipenv
 ---> Using cache
 ---> bfb1bc54efaf
Step 3/9 : WORKDIR /homework
 ---> Using cache
 ---> fa0cfd0eab72
Step 4/9 : COPY ["Pipfile", "Pipfile.lock", "./"]
 ---> Using cache
 ---> 155ba5deb405
Step 5/9 : RUN pipenv install --system --deploy
 ---> Using cache
 ---> f0a76b455a8f
Step 6/9 : COPY ["serve_model.py", "model1.bin", "dv.bin", "./"]
 ---> 8fcc85d931e7
Step 7/9 : EXPOSE 5000
 ---> Running in efc117ea0e18
Removing intermediate container efc117ea0e18
 ---> 862a38ae3c58
Step 8/9 : ENTRYPOINT ["python3"]
 ---> Running in 0bb1d59b23c0
Removing intermediate container 0bb1d59b23c0
 ---> 4391c1dd4630
Step 9/9 : CMD [ "serve_model.py", "--host=0.0.0.0"]
 ---> Running in 674762ae264d
Removing intermediate container 674762ae264d
 ---> 2bb2e29dac7e
Successfully built 2bb2e29dac7e
Successfully tagged svizor/zoomcamp-model:predict


In [62]:
!docker run -p 5000:5000 svizor/zoomcamp-model:predict

 * Serving Flask app 'credit_card'
 * Debug mode: on
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:5000
 * Running on http://172.17.0.2:5000
Press CTRL+C to quit
 * Restarting with stat
 * Debugger is active!
 * Debugger PIN: 722-257-279
172.17.0.1 - - [11/Oct/2022 11:33:02] "GET /predict HTTP/1.1" 405 -
172.17.0.1 - - [11/Oct/2022 11:33:02] "GET /favicon.ico HTTP/1.1" 404 -
172.17.0.1 - - [11/Oct/2022 11:33:07] "GET / HTTP/1.1" 404 -
^C


In [66]:
!docker ps

CONTAINER ID   IMAGE                           COMMAND                  CREATED          STATUS          PORTS                                       NAMES
12276ed87aa8   svizor/zoomcamp-model:predict   "python3 serve_model…"   11 seconds ago   Up 10 seconds   0.0.0.0:5000->5000/tcp, :::5000->5000/tcp   admiring_kirch


## Question 6

Let's run your docker container!

After running it, score this client once again:

```python
url = "YOUR_URL"
client = {"reports": 0, "share": 0.245, "expenditure": 3.438, "owner": "yes"}
requests.post(url, json=client).json()
```

What's the probability that this client will get a credit card now?

* 0.289
* 0.502
* 0.769
* __0.972__ - Closest to 0.928

In [67]:
import requests
url = "http://127.0.0.1:5000/predict"
client = {"reports": 0, "share": 0.245, "expenditure": 3.438, "owner": "yes"}
response = requests.post(url, json=client).json()
print(response)

{'card_approved_probability': 0.9282218018527452}


## Submit the results

* Submit your results here: https://forms.gle/jU2we8f9WeLgX3qa6
* You can submit your solution multiple times. In this case, only the last submission will be used 
* If your answer doesn't match options exactly, select the closest one


## Deadline

The deadline for submitting is **10 October 2022 (Monday), 23:00 CEST (Berlin time)**. 

After that, the form will be closed.